From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: gellert <gellert.hegyi@around.co>
Date: Tue, 30 Jun 2020 18:57:40 +0200
Subject: sets option to not start exclusion in preview

This patch sets a flag that will prevent window exclusion while in preview when doing screensharing

Signed-off-by: gellert <gellert.hegyi@around.co>

diff --git a/modules/desktop_capture/desktop_and_cursor_composer.cc b/modules/desktop_capture/desktop_and_cursor_composer.cc
index 7be8982abeaa9d14d031e761868e68e225ee717c..400cf19ed863430f4f8499978b71236a6ad045e7 100644
--- a/modules/desktop_capture/desktop_and_cursor_composer.cc
+++ b/modules/desktop_capture/desktop_and_cursor_composer.cc
@@ -169,6 +169,7 @@ void DesktopAndCursorComposer::Start(DesktopCapturer::Callback* callback) {
   callback_ = callback;
   if (mouse_monitor_)
     mouse_monitor_->Init(this, MouseCursorMonitor::SHAPE_AND_POSITION);
+  desktop_capturer_->should_exclude_ = should_exclude_;
   desktop_capturer_->Start(this);
 }
 
diff --git a/modules/desktop_capture/desktop_capturer.h b/modules/desktop_capture/desktop_capturer.h
index 8a65e5759cc927c8cf53f874cb604403df4aa26e..6617c69e351bc7fd1e22cb8bc53c8b19dbf4c0f9 100644
--- a/modules/desktop_capture/desktop_capturer.h
+++ b/modules/desktop_capture/desktop_capturer.h
@@ -80,6 +80,7 @@ class RTC_EXPORT DesktopCapturer {
   // Called at the beginning of a capturing session. `callback` must remain
   // valid until capturer is destroyed.
   virtual void Start(Callback* callback) = 0;
+  bool should_exclude_ = false;
 
   // Sets SharedMemoryFactory that will be used to create buffers for the
   // captured frames. The factory can be invoked on a thread other than the one
diff --git a/modules/desktop_capture/mac/screen_capturer_mac.mm b/modules/desktop_capture/mac/screen_capturer_mac.mm
index 6abc0b61c8038a6dfa777cdf7c9f951355fe20ee..4f1d0b83ace88bf4f191530f8fc44eefbd498c3a 100644
--- a/modules/desktop_capture/mac/screen_capturer_mac.mm
+++ b/modules/desktop_capture/mac/screen_capturer_mac.mm
@@ -370,16 +370,18 @@ DesktopRect GetExcludedWindowPixelBounds(CGWindowID window, float dip_to_pixel_s
     // Translate the region to be copied into display-relative coordinates.
     copy_region.Translate(-display_bounds.left(), -display_bounds.top());
 
-    float window_exclude_capture_fps = 5.0f;
-    float delta = 1000.0f / window_exclude_capture_fps;
-    if (last_exclude_capture_ms_ == 0 || rtc::TimeSince(last_exclude_capture_ms_) > delta) {
-      last_exclude_capture_ms_ = rtc::SystemTimeMillis();
-      excluded_images_.clear();
-      excluded_window_bounds_list_.clear();
-      excluded_image_data_list_.clear();
+    if (should_exclude_) {
+      float window_exclude_capture_fps = 5.0f;
+      float delta = 1000.0f / window_exclude_capture_fps;
+      if (last_exclude_capture_ms_ == 0 || rtc::TimeSince(last_exclude_capture_ms_) > delta) {
+        last_exclude_capture_ms_ = rtc::SystemTimeMillis();
+        excluded_images_.clear();
+        excluded_window_bounds_list_.clear();
+        excluded_image_data_list_.clear();
+      }
     }
 
-    if (excluded_images_.size() == 0) {
+    if (should_exclude_ && excluded_images_.size() == 0) {
       for(std::vector<CGWindowID>::size_type i = 0; i < excluded_windows_.size(); i++) {
         CGWindowID excluded_window = excluded_windows_[i];
 
@@ -431,35 +433,37 @@ DesktopRect GetExcludedWindowPixelBounds(CGWindowID window, float dip_to_pixel_s
                it.rect());
     }
 
-    for(std::vector<DesktopRect>::size_type i = 0; i < excluded_window_bounds_list_.size(); i++) {
-      DesktopRect excluded_window_bounds = excluded_window_bounds_list_[i];
-      rtc::ScopedCFTypeRef<CGImageRef> excluded_image = excluded_images_[i];
-      rtc::ScopedCFTypeRef<CFDataRef> excluded_image_data = excluded_image_data_list_[i];
+    if (should_exclude_) {
+      for(std::vector<DesktopRect>::size_type i = 0; i < excluded_window_bounds_list_.size(); i++) {
+        DesktopRect excluded_window_bounds = excluded_window_bounds_list_[i];
+        rtc::ScopedCFTypeRef<CGImageRef> excluded_image = excluded_images_[i];
+        rtc::ScopedCFTypeRef<CFDataRef> excluded_image_data = excluded_image_data_list_[i];
 
       if (!excluded_image_data.get()) {
-        continue;
-      }
-      RTC_DCHECK(excluded_image_data);
-      display_base_address = CFDataGetBytePtr(excluded_image_data.get());
-      src_bytes_per_row = CGImageGetBytesPerRow(excluded_image.get());
-
-      // Translate the bounds relative to the desktop, because `frame` data
-      // starts from the desktop top-left corner.
-      DesktopRect window_bounds_relative_to_desktop(excluded_window_bounds);
-      window_bounds_relative_to_desktop.Translate(-screen_pixel_bounds_.left(),
-                                                  -screen_pixel_bounds_.top());
-
-      DesktopRect rect_to_copy = DesktopRect::MakeSize(excluded_window_bounds.size());
-      rect_to_copy.IntersectWith(DesktopRect::MakeWH(CGImageGetWidth(excluded_image.get()),
-                                                     CGImageGetHeight(excluded_image.get())));
-
-      if (CGImageGetBitsPerPixel(excluded_image.get()) / 8 == DesktopFrame::kBytesPerPixel) {
-        CopyRect(display_base_address,
-                 src_bytes_per_row,
-                 frame.GetFrameDataAtPos(window_bounds_relative_to_desktop.top_left()),
-                 frame.stride(),
-                 DesktopFrame::kBytesPerPixel,
-                 rect_to_copy);
+          continue;
+        }
+        RTC_DCHECK(excluded_image_data);
+        display_base_address = CFDataGetBytePtr(excluded_image_data.get());
+        src_bytes_per_row = CGImageGetBytesPerRow(excluded_image.get());
+
+        // Translate the bounds relative to the desktop, because `frame` data
+        // starts from the desktop top-left corner.
+        DesktopRect window_bounds_relative_to_desktop(excluded_window_bounds);
+        window_bounds_relative_to_desktop.Translate(-screen_pixel_bounds_.left(),
+                                                    -screen_pixel_bounds_.top());
+
+        DesktopRect rect_to_copy = DesktopRect::MakeSize(excluded_window_bounds.size());
+        rect_to_copy.IntersectWith(DesktopRect::MakeWH(CGImageGetWidth(excluded_image.get()),
+                                                      CGImageGetHeight(excluded_image.get())));
+
+        if (CGImageGetBitsPerPixel(excluded_image.get()) / 8 == DesktopFrame::kBytesPerPixel) {
+          CopyRect(display_base_address,
+                  src_bytes_per_row,
+                  frame.GetFrameDataAtPos(window_bounds_relative_to_desktop.top_left()),
+                  frame.stride(),
+                  DesktopFrame::kBytesPerPixel,
+                  rect_to_copy);
+        }
       }
     }
   }
